pkg/apis/logging/v1/output_types.go

```diff
@@ -0,0 +1,109 @@
+package v1
+
+import (
+	"github.com/openshift/cluster-logging-operator/pkg/apis/logging/v1/outputs"
+	corev1 "k8s.io/api/core/v1"
+)
+
+// Output defines a destination for log messages.
+type Output struct {
+	// Name used to refer to the output from a `pipeline`.
+	//
+	// +required
+	Name string `json:"name"`
+
+	// OutputType is inlined with a required `type` and optional extra configuration.
+	OutputType `json:",inline"`
```

Comment:
i thought the k8s pattern is to have something like
```
Type           OutputType `json:"type"`
OneOfTheTypes  TheType `json:"oneOfTheTypes"`
AnotherType    ...
```

Reply:
Agree.  this seems odd to me... The pattern i have seen elsewhere is as described by @ewolinetz comment though carrying the type in the containing struct may allow easier decisions of its type and the things i can/need to do

Reply:
The two are equivalent from the perspective of the resulting k8s API.  Output is the most complex and interesting type in the package so to me it makes sense for it to be its own type. If we want to re-use it in future (e.g. for a free-standing shared Output API object) then this also avoids duplication. However, I can inline it for now if that's what folks are more comfortable with, and we can reconsider if/when we need to re-use it. Thoughts?

Reply:
Also I think a separate type clarifies that the "type" field and the following "optional" fields are in fact a single "union" object, and facilitates writing methods that treat it as such - you can of course glom such methods onto the containing object but to me it feels less clear.

Reply:
> The two are equivalent from the perspective of the resulting k8s API. Output is the most complex and interesting type in the package so to me it makes sense for it to be its own type. If we want to re-use it in future (e.g. for a free-standing shared Output API object) then this also avoids duplication. However, I can inline it for now if that's what folks are more comfortable with, and we can reconsider if/when we need to re-use it. Thoughts?

Can you provide a yaml sample to what this looks like?  I can't seem to wrap my brain around the assembly of the golang structs.  Keeping Output as a separate struct makes sense to me given our intention to maybe roll it out in future as standalone

Reply:
YAML is the same either way, the difference only exists in the Go code, to group the fields that are part of the "union" separately from the rest of the Output struct. They don't have to be grouped in the YAML:
```
output:
  type: syslog
  url: tcp://myhost:1234
  syslog:
     severity: critical
```